datafusion/core/src/execution/memory_manager/mod.rs

```diff
@@ -17,419 +17,187 @@
+/// register their memory requests with the MemoryManager which then tracks the total
+///  memory that has been allocated across all such nodes.
+///
+/// The associated [`MemoryPool`] determines how to respond to memory allocation
+/// requests, and any associated fairness control
+#[derive(Debug)]
+pub struct MemoryManager {
+    pool: Arc<dyn MemoryPool>,
 }
 
-/// Type of the memory consumer
-pub enum ConsumerType {
-    /// consumers that can grow its memory usage by requesting more from the memory manager or
-    /// shrinks its memory usage when we can no more assign available memory to it.
-    /// Examples are spillable sorter, spillable hashmap, etc.
-    Requesting,
-    /// consumers that are not spillable, counting in for only tracking purpose.
-    Tracking,
-}
+impl MemoryManager {
+    /// Create new memory manager based on the configuration
+    pub fn new(pool: Arc<dyn MemoryPool>) -> Self {
+        Self { pool }
+    }
 
-#[derive(Clone, Debug, Hash, Eq, PartialEq)]
-/// Id that uniquely identifies a Memory Consumer
-pub struct MemoryConsumerId {
-    /// partition the consumer belongs to
-    pub partition_id: usize,
-    /// unique id
-    pub id: usize,
-}
+    /// Returns the number of allocated bytes
+    ///
+    /// Note: this can exceed the pool size as a result of [`MemoryManager::allocate`]
+    pub fn allocated(&self) -> usize {
+        self.pool.allocated()
+    }
 
-impl MemoryConsumerId {
-    /// Auto incremented new Id
-    pub fn new(partition_id: usize) -> Self {
-        let id = next_id();
-        Self { partition_id, id }
+    /// Returns a new empty allocation identified by `name`
+    pub fn new_allocation(&self, name: String) -> TrackedAllocation {
+        self.new_allocation_with_options(AllocationOptions::new(name))
     }
-}
 
-impl Display for MemoryConsumerId {
-    fn fmt(&self, f: &mut Formatter) -> fmt::Result {
-        write!(f, "{}:{}", self.partition_id, self.id)
+    /// Returns a new empty allocation with the provided [`AllocationOptions`]
+    pub fn new_allocation_with_options(
+        &self,
+        options: AllocationOptions,
+    ) -> TrackedAllocation {
+        TrackedAllocation::new_empty(options, Arc::clone(&self.pool))
     }
 }
 
-#[async_trait]
-/// A memory consumer that either takes up memory (of type `ConsumerType::Tracking`)
-/// or grows/shrinks memory usage based on available memory (of type `ConsumerType::Requesting`).
-pub trait MemoryConsumer: Send + Sync {
-    /// Display name of the consumer
-    fn name(&self) -> String;
-
-    /// Unique id of the consumer
-    fn id(&self) -> &MemoryConsumerId;
-
-    /// Ptr to MemoryManager
-    fn memory_manager(&self) -> Arc<MemoryManager>;
-
-    /// Partition that the consumer belongs to
-    fn partition_id(&self) -> usize {
-        self.id().partition_id
-    }
+/// Options associated with a [`TrackedAllocation`]
+#[derive(Debug)]
+pub struct AllocationOptions {
```

Comment:
This is designed to be an extension point, so we can potentially add more allocation options and correspondingly more sophisticated MemoryPool.